Catch panics/unwinding in destruction of TLS values

flba-eb opened the pull request:
`destroy_value` is/can be called from C code (libc). Unwinding from Rust to C code is undefined behavior, which is why unwinding is caught here.

This problem caused an infinite loop inside the unwinding code when running `src/test/ui/threads-sendsync/issue-24313.rs` on a tier 3 target (QNX/Neutrino) on aarch64.

See also https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Infinite.20unwinding.20bug.

Nilstrieb commented on library/std/src/thread/local.rs:
```diff
@@ -1137,12 +1137,20 @@ pub mod os {
+        // `always_abort` is called, as the drop function of the value can panic
+        // itself, causing an unwinding to happen outside the `catch_unwind`
+        // call.
+        if let Err(_err) = crate::panic::catch_unwind(|| unsafe {
```
This doesn't guard against `panic_any(DropBomb)`, which we may or may not care about here.

bjorn3 replied:
The DropBomb did get dropped after the aborting panic below, or in other words it will never be dropped at all.

flba-eb replied:
Resolving conversation as I think we have an agreement here that this is not an issue.

riking commented on library/std/src/thread/local.rs:
```diff
@@ -1137,12 +1137,20 @@ pub mod os {
             key.os.set(ptr::null_mut());
+        }) {
+            crate::panic::always_abort();
+            panic!("destruction of a thread local value");
```
slightly more descriptive: "panic during destruction..."

flba-eb replied:
The two lines have been replaced by `rtabort!("destructor panicked");` -- is this okay?

flba-eb committed:
Catch panics in destruction of TLS values

`destroy_value` is/can be called from C code (libc). Unwinding
from Rust to C code is undefined behavior, which is why unwinding
is caught here.

flba-eb commented on library/std/src/thread/local.rs:
```diff
@@ -1028,10 +1025,16 @@ pub mod fast {
+        //
+        // Wrap the call in a catch to ensure unwinding is caught in the event
+        // a panic takes place in a destructor.
+        if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
+            let ptr = ptr as *mut Key<T>;
```
Is `panic::AssertUnwindSafe` correct here? Is it allowed to move `let ptr = ` down to limit its scope?

m-ou-se replied:
You only have to put the `drop()` line in the catch_unwind.

flba-eb replied:
@m-ou-se Unfortunately the dereferencing of raw pointers needs to be done inside the `unsafe` block and it cannot be done in a separate `unsafe` block as we need the variable `value`. I've restored the position of `let ptr = ...` line as it neither requires `catch_unwind` nor `unsafe`. Is this what you had in mind?

m-ou-se commented on library/std/src/thread/local.rs:
```diff
@@ -1028,10 +1025,16 @@ pub mod fast {
             (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
             drop(value);
+        })) {
+            rtabort!("destructor panicked");
```
Can you change the panic message to "thread local panicked on drop"? Then it matches the message of thread results:

https://github.com/rust-lang/rust/blob/6a20f7df5755d8c6b68110d2d0391a7b03268e77/library/std/src/thread/mod.rs#L1424-L1428

flba-eb replied:
Great hint, thank & done

flba-eb commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit 04a6f22f7950daa680594353dbd03f5cb515aef5 has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se 04a6f22f7950daa680594353dbd03f5cb515aef5 -->
<!-- homu: {"type":"Approved","sha":"04a6f22f7950daa680594353dbd03f5cb515aef5","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 04a6f22f7950daa680594353dbd03f5cb515aef5 with merge 7c991868c60a4afc1ee6334b912ea96061a2c98d...
<!-- homu: {"type":"BuildStarted","head_sha":"04a6f22f7950daa680594353dbd03f5cb515aef5","merge_sha":"7c991868c60a4afc1ee6334b912ea96061a2c98d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3806961326/jobs/6477922362)
Approved by: m-ou-se
Pushing 7c991868c60a4afc1ee6334b912ea96061a2c98d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3806961326/jobs/6477922362"},"merge_sha":"7c991868c60a4afc1ee6334b912ea96061a2c98d"} -->

rust-timer commented:
Finished benchmarking commit (7c991868c60a4afc1ee6334b912ea96061a2c98d): [comparison URL](https://perf.rust-lang.org/compare.html?start=f6cc345be403d481b4518e151218118c2a9eb4bb&end=7c991868c60a4afc1ee6334b912ea96061a2c98d&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f6cc345be403d481b4518e151218118c2a9eb4bb&end=7c991868c60a4afc1ee6334b912ea96061a2c98d&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.3%, 1.9%]    | 11    |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.6%, -0.2%]  | 41    |
| Improvements ✅ <br /> (secondary) | -1.8% | [-14.6%, -0.2%] | 31    |
| All ❌✅ (primary)                 | -0.1% | [-0.6%, 1.9%]   | 52    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f6cc345be403d481b4518e151218118c2a9eb4bb&end=7c991868c60a4afc1ee6334b912ea96061a2c98d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [0.4%, 3.4%]   | 8     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.0% | [-8.1%, -1.2%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-8.1%, 3.4%]  | 12    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f6cc345be403d481b4518e151218118c2a9eb4bb&end=7c991868c60a4afc1ee6334b912ea96061a2c98d&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [2.2%, 2.7%]    | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -     | -               | 0     |
| Improvements ✅ <br /> (secondary) | -5.5% | [-14.9%, -2.3%] | 11    |
| All ❌✅ (primary)                 | 2.4%  | [2.2%, 2.7%]    | 2     |
</details>

rylev commented:
@m-ou-se @flba-eb I'm not sure I understand where this perf regression is coming from. This does not look like noise as the impact is pretty broad and impacts some of the more stable benchmarks. Some of the benchmarks are showing increased activity in LLVM which might be caused by extra code gen, but this is not the case in all regressions and it certainly doesn't explain the improvements. 

Any thoughts?

flba-eb commented:
> @m-ou-se @flba-eb I'm not sure I understand where this perf regression is coming from. This does not look like noise as the impact is pretty broad and impacts some of the more stable benchmarks. Some of the benchmarks are showing increased activity in LLVM which might be caused by extra code gen, but this is not the case in all regressions and it certainly doesn't explain the improvements. 
> 
> Any thoughts?

Good question, @rylev . I don't know how CatchUnwind is implemented but I guess it adds a special symbol on the stack which let's the unwinding stop early. On qnx it stops a loop which would run endlessly otherwise (when drop panics) -- I could imagine that on other systems unwinding have a similar issue but have some safety counter to also stop after a high number of e.g. checked frames. In such a case unwinding would stop earlier now (at CatchUnwind instead after reaching some number). But this is really some very wild guess... It also would only make sense if the improved test cases have drop calls that panic.
